front/lib/actions/mcp_internal_actions/servers/schedules_management.ts

```diff
@@ -0,0 +1,502 @@
+import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
+import { z } from "zod";
+
+import { MCPError } from "@app/lib/actions/mcp_errors";
+import { makeInternalMCPServer } from "@app/lib/actions/mcp_internal_actions/utils";
+import { withToolLogging } from "@app/lib/actions/mcp_internal_actions/wrappers";
+import type { AgentLoopContextType } from "@app/lib/actions/types";
+import { generateCronRule } from "@app/lib/api/assistant/configuration/triggers";
+import type { Authenticator } from "@app/lib/auth";
+import { TriggerResource } from "@app/lib/resources/trigger_resource";
+import { getStatsDClient } from "@app/lib/utils/statsd";
+import logger from "@app/logger/logger";
+import type { Result } from "@app/types";
+import { Err, Ok } from "@app/types";
+import type { AgentConfigurationType } from "@app/types/assistant/agent";
+import { isUserMessageType } from "@app/types/assistant/conversation";
+
+function renderSchedule(schedule: TriggerResource): string {
+  const config = schedule.configuration;
+  const scheduleInfo =
+    "cron" in config
+      ? `${schedule.naturalLanguageDescription ?? config.cron} (${config.timezone})`
+      : "";
+  const lines = [
+    `- **${schedule.name}** (ID: ${schedule.sId()})`,
+    `  Schedule: ${scheduleInfo}`,
+  ];
+  if (schedule.customPrompt) {
+    lines.push(`  Prompt: ${schedule.customPrompt}`);
+  }
+  lines.push(`  Enabled: ${schedule.enabled ? "Yes" : "No"}`);
+  return lines.join("\n");
+}
+
+function getToolContext(
+  auth: Authenticator,
+  agentLoopContext?: AgentLoopContextType
+): Result<
+  {
+    userId: number;
+    workspaceId: string;
+    agentConfiguration: AgentConfigurationType;
+  },
+  MCPError
+> {
+  const user = auth.user();
+  const workspace = auth.workspace();
+  if (!user) {
+    return new Err(new MCPError("User not found"));
+  }
+  if (!workspace) {
+    return new Err(new MCPError("Workspace not found"));
+  }
+  if (!agentLoopContext?.runContext) {
+    return new Err(new MCPError("Agent context is required"));
+  }
+  return new Ok({
+    userId: user.id,
+    workspaceId: workspace.sId,
+    agentConfiguration: agentLoopContext.runContext.agentConfiguration,
+  });
+}
```

Comment:
we don't actually have this getToolContext pattern anywhere else in tools, + other tools are using getNonNullableWorkspace / getNonNullableUser

if (!agentLoopContext?.runContext) is fine but idk if worth to have a full method just for that